gems/quilt_rails/lib/quilt_rails/monkey_patches/active_support_reloader.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
```

Comment:
@matthewd can you see if this use case makes sense? I'd expect the `permit_concurrent_loads` to release the lock and allow other threads to acquire the lock.

Reply:
It releases the lock.  But if you have:
- ControllerA
  - Unloads classes
  - Blocking while waiting for a response from a Node server
- ControllerB
  - Attempts to unload classes because @should_reload is still `true`
  - Deadlocks on ControllerA's read lock

Reply:
Hmmm. It seems the app is actually blocked on the `unload` lock.

<details>
<summary>Deadlock log</summary>

```
Thread 0 [0x3fed4d2b3980 sleep]  No lock (yielded share)
  Waiting in start_exclusive to "unload"
  may be pre-empted for: "load", "unload"
  blocked by: 2
  blocking: 1

/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `sleep'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:121:in `wait_while'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:221:in `wait_for'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:83:in `block (2 levels) in start_exclusive'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:82:in `block in start_exclusive'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:77:in `start_exclusive'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:25:in `start_unloading'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/reloader.rb:101:in `require_unload_lock!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/reloader.rb:120:in `class_unload!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/application/finisher.rb:177:in `block (2 levels) in <module:Finisher>'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:426:in `instance_exec'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:426:in `block in make_lambda'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:198:in `block (2 levels) in halting'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:606:in `block (2 levels) in default_terminator'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:605:in `catch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:605:in `block in default_terminator'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:199:in `block in halting'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `block in invoke_before'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `each'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `invoke_before'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:131:in `run_callbacks'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:110:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/reloader.rb:115:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:72:in `block in run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:69:in `tap'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:69:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/reloader.rb:64:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_error_adapter.rb:12:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/bugsnag-6.12.0/lib/bugsnag/integrations/rack.rb:49:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:135:in `call_app'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:22:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:20:in `catch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:20:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:38:in `call_app'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:71:in `block in tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:28:in `tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:71:in `tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/sprockets-rails-3.2.1/lib/sprockets/rails/quiet_assets.rb:13:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_adapter.rb:40:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing.rb:83:in `trace'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_adapter.rb:39:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rails_adapter.rb:61:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/method_override.rb:22:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/runtime.rb:22:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/static.rb:127:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/sendfile.rb:111:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/debug_locks.rb:41:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/ssl.rb:74:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/raindrops_middleware.rb:21:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/request_metrics_middleware.rb:47:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/engine.rb:524:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/configuration.rb:228:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:664:in `handle_request'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:467:in `process_client'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:328:in `block in run'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'


---


Thread 1 [0x3fed4d2b3ae8 sleep]  No lock
  Waiting in start_sharing
  blocked by: 0

/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `sleep'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:109:in `wait'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:121:in `wait_while'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:221:in `wait_for'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:125:in `block in start_sharing'
/opt/rubies/2.5.3/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:115:in `start_sharing'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:33:in `start_running'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/application/finisher.rb:94:in `run'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:27:in `before'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:426:in `block in make_lambda'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:198:in `block (2 levels) in halting'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:606:in `block (2 levels) in default_terminator'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:605:in `catch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:605:in `block in default_terminator'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:199:in `block in halting'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `block in invoke_before'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `each'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:513:in `invoke_before'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:131:in `run_callbacks'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:110:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:72:in `block in run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:69:in `tap'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/execution_wrapper.rb:69:in `run!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/static.rb:127:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/sendfile.rb:111:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/debug_locks.rb:41:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/ssl.rb:74:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/raindrops_middleware.rb:21:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/request_metrics_middleware.rb:47:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/engine.rb:524:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/configuration.rb:228:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:664:in `handle_request'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:467:in `process_client'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:328:in `block in run'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'


---


Thread 2 [0x3fed4d2b3cdc sleep]  No lock (yielded share)
  blocking: 0

/opt/rubies/2.5.3/lib/ruby/2.5.0/net/protocol.rb:181:in `wait_readable'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/protocol.rb:181:in `rbuf_fill'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/protocol.rb:157:in `readuntil'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/protocol.rb:167:in `readline'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http/response.rb:40:in `read_status_line'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http/response.rb:29:in `read_new'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:1494:in `block in transport_request'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:1491:in `catch'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:1491:in `transport_request'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:1464:in `request'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rails-reverse-proxy-0.9.1/lib/reverse_proxy/client.rb:92:in `block in request'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:910:in `start'
/opt/rubies/2.5.3/lib/ruby/2.5.0/net/http.rb:609:in `start'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rails-reverse-proxy-0.9.1/lib/reverse_proxy/client.rb:90:in `request'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rails-reverse-proxy-0.9.1/lib/reverse_proxy/controller.rb:59:in `reverse_proxy'
/Users/rafaelfranca/src/shopify/quilt/gems/quilt_rails/lib/quilt_rails/react_renderable.rb:29:in `block in proxy'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/Users/rafaelfranca/src/shopify/quilt/gems/quilt_rails/lib/quilt_rails/react_renderable.rb:28:in `proxy'
/Users/rafaelfranca/src/shopify/quilt/gems/quilt_rails/lib/quilt_rails/react_renderable.rb:14:in `block in render_react'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_security_base-2.0.1/lib/shopify_security_base/patches/net_http/http_host_restriction.rb:20:in `whitelist'
/Users/rafaelfranca/src/shopify/quilt/gems/quilt_rails/lib/quilt_rails/react_renderable.rb:13:in `render_react'
/Users/rafaelfranca/src/shopify/beta-buddy/app/controllers/ui_controller.rb:7:in `index'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/abstract_controller/base.rb:194:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/rendering.rb:30:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/abstract_controller/callbacks.rb:42:in `block in process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:132:in `run_callbacks'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/abstract_controller/callbacks.rb:41:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/rescue.rb:22:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/notifications.rb:168:in `block in instrument'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/notifications.rb:168:in `instrument'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/instrumentation.rb:32:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activerecord-5.2.3/lib/active_record/railties/controller_runtime.rb:24:in `process_action'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/abstract_controller/base.rb:134:in `process'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionview-5.2.3/lib/action_view/rendering.rb:32:in `process'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal.rb:191:in `dispatch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_controller/metal.rb:252:in `dispatch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/routing/route_set.rb:34:in `serve'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/journey/router.rb:52:in `block in serve'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/journey/router.rb:35:in `each'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/journey/router.rb:35:in `serve'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/routing/route_set.rb:840:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:192:in `call!'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:169:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/omniauth-1.9.0/lib/omniauth/builder.rb:64:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/tempfile_reaper.rb:15:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/etag.rb:25:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/conditional_get.rb:25:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/head.rb:12:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/http/content_security_policy.rb:18:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/session/abstract/id.rb:232:in `context'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/session/abstract/id.rb:226:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/cookies.rb:670:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activerecord-5.2.3/lib/active_record/migration.rb:559:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/callbacks.rb:98:in `run_callbacks'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_error_adapter.rb:12:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/bugsnag-6.12.0/lib/bugsnag/integrations/rack.rb:49:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:135:in `call_app'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:22:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:20:in `catch'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/web-console-3.7.0/lib/web_console/middleware.rb:20:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:38:in `call_app'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:71:in `block in tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:28:in `tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/tagged_logging.rb:71:in `tagged'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/sprockets-rails-3.2.1/lib/sprockets/rails/quiet_assets.rb:13:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_adapter.rb:40:in `block in call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing.rb:83:in `trace'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rack_adapter.rb:39:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify-tracing-0.6.30/lib/shopify/tracing/adapters/rails_adapter.rb:61:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/request_id.rb:27:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/method_override.rb:22:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/runtime.rb:22:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/activesupport-5.2.3/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/static.rb:127:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/rack-2.0.7/lib/rack/sendfile.rb:111:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/debug_locks.rb:41:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/actionpack-5.2.3/lib/action_dispatch/middleware/ssl.rb:74:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/raindrops_middleware.rb:21:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/shopify_metrics-1.1.1/lib/shopify_metrics/middlewares/request_metrics_middleware.rb:47:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/railties-5.2.3/lib/rails/engine.rb:524:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/configuration.rb:228:in `call'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:664:in `handle_request'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:467:in `process_client'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/server.rb:328:in `block in run'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/puma-4.1.1/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
/Users/rafaelfranca/.gem/ruby/2.5.3/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```
</details>

And we don't have a method to allow unload concurrently. Is that unsafe @matthewd?

Reply:
Yeah, it's unsafe to unload in the middle of a block of code, because:

```ruby
one = A.new
method_that_sleeps_while_permitting_unloads
two = A.new
one.class != two.class
```